packages/@react-spectrum/s2/src/Tabs.tsx

```diff
@@ -321,43 +295,301 @@ function TabLine(props: TabLineProps) {
+  let testKey: Key | null = null;
+  if (collection && collection.size > 0) {
+    testKey = collection.getFirstKey();
+
+    let index = 0;
+    while (testKey && index < collection.size) {
+      // We have to check if the item in the collection has a key in disabledKeys or has the isDisabled prop set directly on it
+      if (!disabledKeys.has(testKey) && !collection.getItem(testKey)?.props?.isDisabled) {
+        return false;
+      }
+
+      testKey = collection.getKeyAfter(testKey);
+      index++;
+    }
+    return true;
+  }
+  return false;
+}
+
+let HiddenTabs = function (props: {
+  listRef: RefObject<HTMLDivElement | null>,
+  items: Array<Node<any>>,
+  size?: string,
+  density?: 'compact' | 'regular'
+}) {
+  let {listRef, items, size, density} = props;
+
+  return (
+    <div
+      // @ts-ignore
+      inert="true"
+      ref={listRef}
+      className={style({
+        display: '[inherit]',
+        flexDirection: '[inherit]',
+        gap: '[inherit]',
+        flexWrap: '[inherit]',
+        position: 'absolute',
+        inset: 0,
+        visibility: 'hidden',
+        overflow: 'hidden',
+        opacity: 0
+      })}>
+      {items.map((item) => {
+        // pull off individual props as an allow list, don't want refs or other props getting through
+        return (
+          <div
+            data-hidden-tab
+            style={item.props.UNSAFE_style}
+            key={item.key}
+            className={item.props.className({size, density})}>
+            {item.props.children({size, density})}
+          </div>
+        );
+      })}
+    </div>
+  );
+};
+
+let TabsMenu = (props: {items: Array<Node<any>>, onSelectionChange: TabsProps['onSelectionChange']}) => {
+  let {items} = props;
+  let {density, onSelectionChange, selectedKey, isDisabled, disabledKeys} = useContext(InternalTabsContext);
+  let state = useContext(TabListStateContext);
+  let allKeysDisabled = useMemo(() => {
+    return isAllTabsDisabled(state?.collection, disabledKeys ? new Set(disabledKeys) : new Set());
+  }, [state?.collection, disabledKeys]);
+
+  return (
+    <UNSTABLE_CollectionRendererContext.Provider value={UNSTABLE_DefaultCollectionRenderer}>
+      <div
+        className={style({
+          display: 'flex',
+          alignItems: 'center',
+          height: {
+            density: {
+              compact: 32,
+              regular: 48
+            }
+          }})({density})}>
+        <Picker
+          isDisabled={isDisabled || allKeysDisabled}
+          density={density!}
+          items={items}
+          disabledKeys={disabledKeys}
+          selectedKey={selectedKey}
+          onSelectionChange={onSelectionChange}
+          aria-label={'Tab selector'}>
+          {(item: Node<any>) => {
+            // need to determine the best way to handle icon only -> icon and text
+            // good enough to aria-label the picker item?
```

Comment:
An interesting situation can happen with the icon only tabs: if you collapse https://reactspectrum.blob.core.windows.net/reactspectrum/777f4b761ccd6a3e45cbdd9d8ee4d623ee6e3516/storybook-s2/index.html?path=/story/tabs--icons&providerSwitcher-locale=he-IL, you will never be able to see the full contents of two of the options because the point of collapse for the icon only tabs is too small. This maybe partially mitigated by the work you are doing to support tooltips on collection items, but maybe this means we should only render the icon of the tab even when it is collapsed but display the full text in the menu dropdown like it already does. Alternatively, we could instead just have the text truncate so it doesn't get cut off

Happy for this to be followup and for others to weigh in.

Reply:
I don't quite follow, are you talking about how the picker shows the label and runs off the overflow area? So the picker should only have the icon?
here's the two options I was trying to decide between
![Screenshot 2024-11-29 at 8 50 50 am](https://github.com/user-attachments/assets/327ee45a-1013-443d-b503-a4f2c577c5fc)
![Screenshot 2024-11-29 at 8 53 59 am](https://github.com/user-attachments/assets/5d48c005-bfd9-4568-bff1-f430cbb6aa87)
![Screenshot 2024-11-29 at 8 54 22 am](https://github.com/user-attachments/assets/c8689020-2ab3-421c-98a7-cd94b1a29720)

The reason I went the direction I did was so it would more easily label, but I could change it to have the display none text aria-label the icon possibly.

Reply:
Yep, you basically got the gist of it, essentially the user doesn't have a great way of seeing the selected tab's full text other than opening the dropdown. IMO option 2 that you show above isn't too bad for sighted users since the non collapsed tabs already didn't show the tab's text and aria-labeling would work for screen reader users.